crates/fluvio-protocol/src/core/decoder.rs

```diff
@@ -400,6 +401,21 @@ impl DecoderVarInt for Option<Vec<u8>> {
     }
 }
 
+impl<M> Decoder for Arc<M>
```

Comment:
Do we need this? This should be done transparently by Arc since it implements Deref.

Reply:
It is not necessary for Arc<R:Encoder> variable because it will be coerced to R that impl Encode, that's correct. But trait bound `Arc<R>: Encoder`needs also for Arc<R:Request> to impl Request.

Reply:
Problem is that this assumes there is one owner since it performs `get_mut` which doesn't seem to make sense for decoding.   Instead, in this case, it's better to have the caller have an explicit conversion, so this can be written as:
```
   let data = Arc::get_mut(self).
   let some = data.decode(....)
```

Reply:
This is done also for `Arc<R:Request>` to impl `Request` because `Request` requires `Encoder` and `Decoder`. 
It is very unlikely that Arc will be shared at the time of decoding. Usually, decoding happens right after instantiation. But if it happens, we provide error, so it can be identified and fixed.

Reply:
What's the use case for decoding directly from Arc? This seems to be an unsafe operation.  Either we should wrap with a Lock or another kind of atomic operation such that it's safe to access underlying data.

Reply:
For example, we have a method with the following signature:
```
async fn send_receive<R>(&self, request: R) where R: Request {...}
```
and we have some request:
```
struct ProduceRequest<R> where R: Encoder + Decoder + Default + Debug
```
Now with that changes, we can do:
```
let request = Arc::new(ProduceRequest{...});
socker.send_receive(request.clone())
```
It becomes useful in context with retries. We don't need to clone request, just clone arc reference:
```
let request = Arc::new(ProduceRequest{...});
retry(..., || socker.send_receive(request.clone())).await;
```

Reply:
It is not unsafe. If the error occurs, it is just IO error that we return, nothing special. But, if it happens it is the developer's mistake, nothing external can incur it. So, we will just fix it.